src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs

```diff
@@ -26,6 +27,38 @@ internal ReflectionJsonTypeInfo(JsonConverter converter, JsonSerializerOptions o
             if (PropertyInfoForTypeInfo.ConverterStrategy == ConverterStrategy.Object)
             {
                 AddPropertiesAndParametersUsingReflection();
+
+#if NET7_0_OR_GREATER
+                // Compiler adds RequiredMemberAttribute to type if any of the members is marked with 'required' keyword.
+                // SetsRequiredMembersAttribute means that all required members are assigned by constructor and therefore there is no enforcement
+                bool shouldCheckMembersForRequiredMemberAttribute =
```

Comment:
Does C# 11 not support `required` members in older TFMs? It currently fails to compile in `net6.0` with the compiler version installed on my local machine:
```
error CS0656: Missing compiler required member 'System.Runtime.CompilerServices.RequiredMemberAttribute..ctor'
error CS0656: Missing compiler required member 'System.Runtime.CompilerServices.CompilerFeatureRequiredAttribute..ctor'
```
But I wouldn't be surprised if future releases of Roslyn might attempt to polyfill these attributes in older TFMs?

@333fred thoughts?

Reply:
> But I wouldn't be surprised if future releases of Roslyn might attempt to polyfill these attributes in older TFMs?

Roslyn will never attempt to polyfill these attributes. _Users_ might manually include them, but we will not.

Reply:
Also, to be clear:

> Does C# 11 not support required members in older TFMs? 

Since C# 8, our policy has been that new language versions are not officially supported on older TFMs. Users can and do enable newer language versions on older TFMs, but may run into errors like above. They often do manually polyfill in the missing types or depend on libraries that do it for them. But these are unsupported configurations, and various parts of certain features will never work, no matter how they're polyfilled (basically anything that gets a flag in `System.Runtime.CompilerServices.RuntimeFeature`.

Reply:
> Roslyn will never attempt to polyfill these attributes.

Interesting. I kind of assumed the compiler was polyfilling `IsExternalInit` for `netstandard2.0` but turns out we were doing it manually.

> Users might manually include them, but we will not.

That suggests to me that we might want to look up `RequiredMemberAttribute` dynamically in older TFMs.

Reply:
> That suggests to me that we might want to look up RequiredMemberAttribute dynamically in older TFMs.

Yes, I think that's probably true. You may also want to consider the scenario where there are many `RequiredMemberAttribute` definitions: user defines their own, and their model uses a type from the library that defines one themselves. Users commonly generate these attributes as `internal` to avoid conflicts with other libraries.

Reply:
@krwq it looks like we might want to add helper methods in `ReflectionExtensions.cs` that specifically look up `RequiredMemberAttribute`. For `net7.0` targets it would simply look for the known type in the shared framework, whereas for anything older we'd be calling `memberInfo.GetCustomAttributes()` looking for anything whose full name matches that of `RequiredMemberAttribute`. 

We can then remove the build conditionals from this location and integrate this logic with all other attribute resolution inside `AddPropertiesAndParametersUsingReflection`.

Reply:
I'll leave this change off for now and file an issue

Reply:
I'll investigate this further as I missed your last answer when I responded.

Reply:
> @krwq it looks like we might want to add helper methods in ReflectionExtensions.cs that specifically look up RequiredMemberAttribute/CompilerFeatureRequiredAttribute. For net7.0 targets it would simply look for the known type in the shared framework, whereas for anything older we'd be calling memberInfo.GetCustomAttributes() looking for anything whose full name matches that of RequiredMemberAttribute.

My suggestion would be to _always_ only look at FQN. For example, someone who polyfilled a netstandard2.0 library could be depended on by an app that is targeting net7.0.

Reply:
I've updated the PR and took liberty of providing that as an example in the comment